dev/src/transaction.ts

```diff
@@ -622,17 +619,124 @@ export class Transaction implements firestore.Transaction {
   }
 
   /**
-   * Delays further operations based on the provided error.
-   *
-   * @private
-   * @internal
-   * @return A Promise that resolves after the delay expired.
+   * Given a function that performs a read operation, ensures that the first one
+   * is provided with new transaction options and all subsequent ones are queued
+   * upon the resulting transaction ID.
    */
-  private async maybeBackoff(error?: GoogleError): Promise<void> {
-    if ((error?.code as number | undefined) === StatusCode.RESOURCE_EXHAUSTED) {
-      this._backoff.resetToMax();
+  private withLazyStartedTransaction<TParam, TResult>(
+    param: TParam,
+    resultFn: (
+      this: typeof this,
+      param: TParam,
+      opts: Uint8Array | api.ITransactionOptions | Timestamp
+    ) => Promise<{transaction?: Uint8Array; result: TResult}>
+  ): Promise<TResult> {
+    if (this._transactionIdPromise) {
+      // Simply queue this subsequent read operation after the first read
+      // operation has resolved and we don't expect a transaction ID in the
+      // response because we are not starting a new transaction
+      return this._transactionIdPromise.then(async opts => {
+        const r = await resultFn.call(this, param, opts);
+        return r.result;
+      });
+    } else {
+      if (this._readOnlyReadTime) {
+        // We do not start a transaction for read-only transactions
+        // do not set _prevTransactionId
+        return resultFn
+          .call(this, param, this._readOnlyReadTime)
+          .then(resolveResult);
+      } else {
+        // This is the first read of the transaction so we create the appropriate
+        // options for lazily starting the transaction inside this first read op
+        let opts: api.ITransactionOptions;
+        if (this._writeBatch) {
+          opts = this._prevTransactionId
+            ? {readWrite: {retryTransaction: this._prevTransactionId}}
+            : {readWrite: {}};
+        } else {
+          opts = {readOnly: {}};
+        }
```

Comment:
Not sure how much better this is, but I found this logic could maybe be written clearer. Your call. This is not a deal breaker.

```suggestion
        const opts: api.ITransactionOptions = {};
        if (this._writeBatch) {
          opts.readWrite = this._prevTransactionId
            ? {retryTransaction: this._prevTransactionId}
            : {};
        } else {
          opts.readOnly = {};
        }
```

Reply:
Sure